File: android/app/src/main/java/dev/rahier/pouleparty/util/ProfanityFilter.kt
package dev.rahier.pouleparty.util

/**
 * Simple local profanity filter for nickname validation.
 * Checks against a list of blocked words (FR + EN) with basic leetspeak normalization.
 */
object ProfanityFilter {

    private val blockedWords = setOf(
        // English
        "fuck", "shit", "ass", "asshole", "bitch", "bastard", "dick", "cock",
        "pussy", "cunt", "whore", "slut", "fag", "faggot", "nigger", "nigga",
        "retard", "rape", "rapist", "nazi", "penis", "vagina", "dildo",
        "wanker", "twat", "bollocks", "prick", "motherfucker",
        // French
        "merde", "putain", "salope", "pute", "connard", "connasse", "enculer",
        "encule", "encule", "nique", "ntm", "fdp", "pd", "tapette", "gouine",
        "batard", "batard", "bordel", "bite", "couille", "branleur",
        "branleuse", "tg", "negro", "negre", "pede", "pedale",
    )

    /** Returns `true` when the text contains a blocked word. */
    fun containsProfanity(text: String): Boolean {
        val cleaned = text.lowercase()
            .replace("0", "o")
            .replace("1", "i")
            .replace("3", "e")
            .replace("4", "a")
            .replace("5", "s")
            .replace("@", "a")
            .replace("$", "s")
            // Strip common accents
            .let { java.text.Normalizer.normalize(it, java.text.Normalizer.Form.NFD) }
            .replace(Regex("[\\p{InCombiningDiacriticalMarks}]"), "")

        return blockedWords.any { cleaned.contains(it) }
    }
}

File: ios/PouleParty/App/ProfanityFilter.swift
//
//  ProfanityFilter.swift
//  PouleParty
//

import Foundation

enum ProfanityFilter {

    /// Blocked words (FR + EN). Checked as whole-word substrings (case-insensitive).
    private static let blockedWords: Set<String> = [
        // English
        "fuck", "shit", "ass", "asshole", "bitch", "bastard", "dick", "cock",
        "pussy", "cunt", "whore", "slut", "fag", "faggot", "nigger", "nigga",
        "retard", "rape", "rapist", "nazi", "penis", "vagina", "dildo",
        "wanker", "twat", "bollocks", "prick", "motherfucker",
        // French
        "merde", "putain", "salope", "pute", "connard", "connasse", "enculer",
        "encule", "enculé", "nique", "ntm", "fdp", "pd", "tapette", "gouine",
        "batard", "bâtard", "bordel", "bite", "couille", "branleur",
        "branleuse", "tg", "negro", "nègre", "pédé", "pédale",
    ]

    /// Returns `true` when the nickname contains a blocked word.
    static func containsProfanity(_ text: String) -> Bool {
        let lowered = text.lowercased()
            .folding(options: .diacriticInsensitive, locale: .current)

        // Also strip common leetspeak substitutions
        let cleaned = lowered
            .replacingOccurrences(of: "0", with: "o")
            .replacingOccurrences(of: "1", with: "i")
            .replacingOccurrences(of: "3", with: "e")
            .replacingOccurrences(of: "4", with: "a")
            .replacingOccurrences(of: "5", with: "s")
            .replacingOccurrences(of: "@", with: "a")
            .replacingOccurrences(of: "$", with: "s")

        for word in blockedWords {
            let normalizedWord = word.lowercased()
                .folding(options: .diacriticInsensitive, locale: .current)
            if cleaned.contains(normalizedWord) {
                return true
            }
        }
        return false
    }
}
